plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go

```diff
@@ -0,0 +1,516 @@
+
+	log.Printf("I! Started Cisco MDT service on %s", c.ServiceAddress)
+
+	return nil
+}
+
+// AcceptTCPDialoutClients defines the TCP dialout server main routine
+func (c *CiscoTelemetryMDT) acceptTCPDialoutClients() {
+	// Keep track of all active connections, so we can close them if necessary
+	var mutex sync.Mutex
+	clients := make(map[net.Conn]struct{})
+
+	for c.ctx.Err() == nil {
+		conn, err := c.listener.Accept()
+		if err != nil {
+			if c.ctx.Err() == nil {
+				c.acc.AddError(fmt.Errorf("Failed to accept TCP connection: %v", err))
+			}
+			continue
+		}
+
+		mutex.Lock()
+		clients[conn] = struct{}{}
+		mutex.Unlock()
+
+		// Individual client connection routine
+		c.wg.Add(1)
+		go func() {
+			log.Printf("D! Accepted Cisco MDT TCP dialout connection from %s", conn.RemoteAddr())
+
+			// TCP Dialout telemetry framing header
+			var hdr struct {
+				MsgType       uint16
+				MsgEncap      uint16
+				MsgHdrVersion uint16
+				MsgFlags      uint16
+				MsgLen        uint32
+			}
+
+			var payload bytes.Buffer
+
+			for c.ctx.Err() == nil {
+				// Read and validate dialout telemetry header
+				if err := binary.Read(conn, binary.BigEndian, &hdr); err != nil {
+					if c.ctx.Err() == nil && err != io.EOF {
+						c.acc.AddError(fmt.Errorf("Unable to read dialout header: %v", err))
+					}
+					break
+				}
+
+				maxMsgSize := tcpMaxMsgLen
+				if c.MaxMsgSize > 0 {
+					maxMsgSize = uint32(c.MaxMsgSize)
+				}
+
+				if hdr.MsgLen > maxMsgSize {
+					c.acc.AddError(fmt.Errorf("Dialout packet too long: %v", hdr.MsgLen))
+					break
+				}
+
+				if hdr.MsgFlags != 0 {
+					c.acc.AddError(fmt.Errorf("Invalid dialout flags: %v", hdr.MsgFlags))
+					break
+				}
+
+				// Read and handle telemetry packet
+				payload.Reset()
+				if size, err := payload.ReadFrom(io.LimitReader(conn, int64(hdr.MsgLen))); size != int64(hdr.MsgLen) {
+					if c.ctx.Err() == nil {
+						if err != nil {
+							c.acc.AddError(fmt.Errorf("TCP dialout I/O error: %v", err))
+						} else {
+							c.acc.AddError(fmt.Errorf("TCP dialout premature EOF"))
+						}
+					}
+					break
+				}
+
+				c.handleTelemetry(payload.Bytes())
+			}
+
+			log.Printf("D! Closed Cisco MDT TCP dialout connection from %s", conn.RemoteAddr())
+
+			mutex.Lock()
+			delete(clients, conn)
+			mutex.Unlock()
+
+			conn.Close()
+			c.wg.Done()
+		}()
+	}
+
+	// Close all remaining client connections
+	mutex.Lock()
+	for client := range clients {
+		if err := client.Close(); err != nil {
+			log.Printf("E! Failed to close TCP dialout client: %v", err)
+		}
+	}
+	mutex.Unlock()
+
+	c.listener.Close()
+	c.wg.Done()
+}
+
+// MdtDialout RPC server method for grpc-dialout transport
+func (c *CiscoTelemetryMDT) MdtDialout(stream dialout.GRPCMdtDialout_MdtDialoutServer) error {
+	peer, peerOK := peer.FromContext(stream.Context())
+	if peerOK {
+		log.Printf("D! Accepted Cisco MDT GRPC dialout connection from %s", peer.Addr)
+	}
+
+	for c.ctx.Err() == nil {
+		packet, err := stream.Recv()
```

Comment:
I'm not super familiar with gRPC, but how does this call to Recv get cancelled on shutdown?  Is it interrupted by the context or does it require reading a message?

I'm concerned this will hang if no further messages are received on the stream.

Reply:
You may be right, I have switched to using grpc.Server Stop() which should take care of that:

> Stop stops the gRPC server. It immediately closes all open connections and listeners. It cancels all active RPCs on the server side and the corresponding pending RPCs on the client side will get notified by connection errors.